File: bin/plot_param_heatmap.R
#!/usr/bin/env Rscript
# Combined parameter heatmaps with annotation color bars.
#
# - Rows = parameter combinations (reference x method x subsample_ref)
# - Columns = cell types
# - Faceted by taxonomy level (key)
# - Left-side color bars encode method, reference, and subsample_ref
#
# Usage:
#   Rscript plot_param_heatmap.R \
#     --input rankings_detailed.tsv \
#     --outdir param_heatmaps

suppressPackageStartupMessages({
  library(argparse)
  library(dplyr)
  library(tidyr)
  library(readr)
  library(stringr)
  library(ggplot2)
  library(patchwork)
  library(scales)
})

KEY_ORDER <- c("subclass", "class", "family", "global")
METHOD_COLORS <- c(scvi = "#1f77b4", seurat = "#ff7f0e")
METHOD_NAMES <- c(scvi = "scVI", seurat = "Seurat")

REF_PALETTE <- c(
  "#66c2a5", "#fc8d62", "#8da0cb", "#e78ac3",
  "#a6d854", "#ffd92f", "#e5c494", "#b3b3b3"
)
SUB_PALETTE <- c("#feedde", "#fdae6b", "#e6550d")

build_ref_short_names <- function(references) {
  refs_sorted <- sort(unique(references))
  short <- case_when(
    grepl("10x", refs_sorted) ~ "10x",
    grepl("SMART", refs_sorted) ~ "SMART-Seq",
    grepl("motor", refs_sorted, ignore.case = TRUE) ~ "Motor Ctx",
    grepl("whole", refs_sorted, ignore.case = TRUE) ~ "Whole Ctx",
    TRUE ~ substr(refs_sorted, 1, 15)
  )
  setNames(short, refs_sorted)
}

cluster_order_cols <- function(df_key) {
  mat <- df_key %>%
    select(row_id, label, mean_f1_across_studies) %>%
    pivot_wider(names_from = label, values_from = mean_f1_across_studies)

  if (nrow(mat) == 0) return(character())

  labels <- setdiff(colnames(mat), "row_id")
  if (length(labels) <= 1) return(labels)

  m <- as.matrix(mat[, labels, drop = FALSE])
  keep <- colSums(!is.na(m)) > 0
  labels <- labels[keep]
  m <- m[, keep, drop = FALSE]

  if (length(labels) <= 1) return(labels)

  for (j in seq_len(ncol(m))) {
    col_mean <- mean(m[, j], na.rm = TRUE)
    if (is.nan(col_mean)) col_mean <- 0
    m[is.na(m[, j]), j] <- col_mean
  }

  cor_mat <- suppressWarnings(cor(m, use = "pairwise.complete.obs"))
  cor_mat[is.na(cor_mat)] <- 0
  dist_mat <- as.dist(1 - cor_mat)
  hc <- hclust(dist_mat, method = "average")
  labels[hc$order]
}

main <- function() {
  parser <- argparse::ArgumentParser(
    description = "Combined parameter heatmaps with annotation color bars"
  )
  parser$add_argument("--input", default = "rankings_detailed.tsv",
                      help = "Path to rankings_detailed.tsv")
  parser$add_argument("--outdir", default = "param_heatmaps",
                      help = "Output directory")
  args <- parser$parse_args()

  dir.create(args$outdir, showWarnings = FALSE, recursive = TRUE)

  df <- read_tsv(args$input, show_col_types = FALSE)
  df$key <- tolower(df$key)

  keys <- intersect(KEY_ORDER, unique(df$key))
  if (length(keys) == 0) {
    message("No keys found. Exiting.")
    return()
  }

  # Build short reference names and palettes
  ref_short <- build_ref_short_names(df$reference)
  refs_sorted <- sort(unique(df$reference))
  ref_colors <- setNames(REF_PALETTE[seq_along(refs_sorted)], refs_sorted)

  subs_sorted <- sort(unique(as.character(df$subsample_ref)))
  sub_colors <- setNames(SUB_PALETTE[seq_along(subs_sorted)], subs_sorted)

  # Create a unique row ID per config, sorted by method then reference then subsample
  df <- df %>%
    mutate(
      method = factor(method, levels = names(METHOD_NAMES)),
      subsample_ref = factor(subsample_ref, levels = subs_sorted),
      row_id = paste0(method, ":", reference, ":", subsample_ref)
    )

  row_meta <- df %>%
    distinct(row_id, method, reference, subsample_ref) %>%
    arrange(method, reference, subsample_ref)
  row_levels <- row_meta$row_id

  df <- df %>%
    mutate(row_id = factor(row_id, levels = row_levels))

  # Cluster x-axis (cell type labels) per key
  label_levels <- c()
  for (k in keys) {
    key_df <- df %>% filter(key == k)
    labels_ordered <- cluster_order_cols(key_df)
    label_levels <- c(label_levels, paste(k, labels_ordered, sep = "___"))
  }

  df <- df %>%
    mutate(
      label_key = paste(key, label, sep = "___"),
      label_key = factor(label_key, levels = label_levels),
      key = factor(key, levels = KEY_ORDER)
    )

  # -- Annotation bar data --
  anno <- row_meta %>%
    mutate(
      row_id = factor(row_id, levels = row_levels),
      subsample_ref = as.character(subsample_ref)
    )

  # Shared y-axis theme for annotation strips
  strip_theme <- theme_void() +
    theme(
      plot.margin = margin(0, 0, 0, 0),
      axis.text.y = element_blank(),
      axis.ticks.y = element_blank()
    )

  # Method annotation bar
  p_method <- ggplot(anno, aes(x = 1, y = row_id, fill = method)) +
    geom_tile(color = "white", linewidth = 0.25) +
    scale_fill_manual(values = METHOD_COLORS,
                      labels = METHOD_NAMES,
                      name = "Method") +
    strip_theme +
    labs(x = NULL, y = NULL)

  # Reference annotation bar
  p_ref <- ggplot(anno, aes(x = 1, y = row_id, fill = reference)) +
    geom_tile(color = "white", linewidth = 0.25) +
    scale_fill_manual(values = ref_colors,
                      labels = ref_short,
                      name = "Reference") +
    strip_theme +
    labs(x = NULL, y = NULL)

  # Subsample annotation bar
  p_sub <- ggplot(anno, aes(x = 1, y = row_id, fill = subsample_ref)) +
    geom_tile(color = "white", linewidth = 0.25) +
    scale_fill_manual(values = sub_colors,
                      name = "Subsample\nRef") +
    strip_theme +
    labs(x = NULL, y = NULL)

  # -- Main heatmap --
  p_heat <- ggplot(df, aes(x = label_key, y = row_id,
                            fill = mean_f1_across_studies)) +
    geom_tile(color = "white", linewidth = 0.25) +
    scale_fill_viridis_c(limits = c(0, 1), name = "Mean F1") +
    scale_x_discrete(labels = function(x) sub(".*___", "", x)) +
    facet_grid(. ~ key, scales = "free_x", space = "free_x") +
    labs(x = NULL, y = NULL) +
    theme_minimal(base_size = 12) +
    theme(
      axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 8),
      axis.text.y = element_blank(),
      axis.ticks.y = element_blank(),
      panel.grid = element_blank(),
      strip.text = element_text(face = "bold", size = 11),
      legend.position = "right"
    )

  # -- Assemble with patchwork --
  combined <- p_method + p_ref + p_sub + p_heat +
    plot_layout(widths = c(0.02, 0.02, 0.02, 1),
                guides = "collect") +
    plot_annotation(
      title = "Parameter Performance Heatmaps",
      theme = theme(
        plot.title = element_text(face = "bold", size = 16, hjust = 0)
      )
    ) &
    theme(legend.position = "right")

  outpath <- file.path(args$outdir, "param_heatmaps_combined.png")
  ggsave(outpath, combined, width = 16, height = 8, dpi = 200, bg = "white")
  message("Saved: ", outpath)
}

main()

File: bin/plot_param_heatmap.py
#!/usr/bin/env python3
"""
plot_best_param_heatmap.py

Clustered heatmaps of mean F1 across studies, faceted by taxonomy level (key).
Rows = parameter combinations (reference × method × subsample_ref),
columns = cell types. Row annotation bars encode reference, method, and
subsample_ref; column annotation bars encode n_studies and mean_support.

Usage:
    python plot_best_param_heatmap.py \
        --input rankings_detailed.tsv \
        --outdir param_heatmaps
"""

import argparse
import os
import string

import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import numpy as np
import pandas as pd
import seaborn as sns

import sys
sys.path.insert(0, os.path.dirname(__file__))
from plot_utils import METHOD_COLORS, METHOD_NAMES, KEY_ORDER, set_pub_style


def parse_args():
    parser = argparse.ArgumentParser(
        description="Clustered heatmaps of parameter performance per cell type"
    )
    parser.add_argument(
        "--input",
        default="rankings_detailed.tsv",
        help="Path to rankings_detailed.tsv",
    )
    parser.add_argument(
        "--outdir", default="param_heatmaps",
        help="Output directory for per-key heatmap PNGs",
    )
    return parser.parse_args()


def build_ref_short_names(references):
    """Assign single-letter codes (A, B, C, …) to references, sorted."""
    refs_sorted = sorted(references)
    return {ref: string.ascii_uppercase[i] for i, ref in enumerate(refs_sorted)}


def make_clustermap(key, key_df, ref_short, ref_palette, sub_palette, outdir):
    """Create and save a clustered heatmap for one taxonomy level."""

    # Build concise row labels: "A:seurat:500"
    key_df = key_df.copy()
    key_df["row_label"] = (
        key_df["reference"].map(ref_short)
        + ":" + key_df["method"]
        + ":" + key_df["subsample_ref"].astype(str)
    )

    # Pivot: rows = param combos, cols = cell types
    heatmap_data = key_df.pivot_table(
        index="row_label", columns="label", values="mean_f1_across_studies"
    )

    # Build lookup from row_label → original columns
    combo_lookup = (
        key_df.drop_duplicates("row_label")
        .set_index("row_label")[["reference", "method", "subsample_ref"]]
    )

    # --- Row annotation colors ---
    row_colors = pd.DataFrame(index=heatmap_data.index)
    row_colors["Reference"] = [
        ref_palette[combo_lookup.loc[rl, "reference"]] for rl in heatmap_data.index
    ]
    row_colors["Method"] = [
        METHOD_COLORS.get(combo_lookup.loc[rl, "method"], "#333333")
        for rl in heatmap_data.index
    ]
    row_colors["Subsample"] = [
        sub_palette[combo_lookup.loc[rl, "subsample_ref"]]
        for rl in heatmap_data.index
    ]

    # --- Column annotation colors ---
    # n_studies per cell type (same across param combos for a given label+key)
    label_meta = key_df.groupby("label").agg(
        n_studies=("n_studies", "first"),
        mean_support=("mean_support", "first"),
    )

    n_studies_vals = label_meta["n_studies"].reindex(heatmap_data.columns)
    support_vals = label_meta["mean_support"].reindex(heatmap_data.columns)

    # Normalize for colormaps
    n_max = n_studies_vals.max() if n_studies_vals.max() > 0 else 1
    s_max = support_vals.max() if support_vals.max() > 0 else 1
    cmap_n = plt.cm.Blues
    cmap_s = plt.cm.Oranges

    col_colors = pd.DataFrame(index=heatmap_data.columns)
    col_colors["N Studies"] = [cmap_n(v / n_max) for v in n_studies_vals]
    col_colors["Support"] = [cmap_s(v / s_max) for v in support_vals]

    # --- Fill NaN for clustering (shouldn't happen, but defensive) ---
    mask = heatmap_data.isna()
    heatmap_data_filled = heatmap_data.fillna(0)

    # --- Plot ---
    n_rows, n_cols = heatmap_data.shape
    fig_w = max(8, n_cols * 0.6 + 4)
    fig_h = max(6, n_rows * 0.35 + 3)

    g = sns.clustermap(
        heatmap_data_filled,
        mask=mask.reindex(index=heatmap_data_filled.index, columns=heatmap_data_filled.columns),
        cmap="viridis",
        row_colors=row_colors,
        col_colors=col_colors,
        linewidths=0.3,
        linecolor="gray",
        figsize=(fig_w, fig_h),
        cbar_kws={"label": "Mean F1 Across Studies"},
        dendrogram_ratio=(0.12, 0.12),
        xticklabels=True,
        yticklabels=False,
    )

    g.ax_heatmap.set_xlabel("Cell Type", fontsize=14)
    g.ax_heatmap.set_ylabel("")
    g.ax_heatmap.tick_params(axis="x", labelsize=12, rotation=90)

    # Shrink annotation bar labels ("N Studies", "Support", "Reference", etc.)
    g.ax_col_colors.tick_params(axis="y", labelsize=8)
    g.ax_row_colors.tick_params(axis="x", labelsize=8)

    # Title
    g.fig.suptitle(f"Parameter Performance — {key}", fontsize=14, fontweight="bold", y=1.02)

    # --- Move colorbar to right side ---
    g.cax.set_visible(False)
    hm_pos = g.ax_heatmap.get_position()
    cbar_ax = g.fig.add_axes([
        hm_pos.x1 + 0.02,
        hm_pos.y0 + hm_pos.height * 0.3,
        0.015,
        hm_pos.height * 0.4,
    ])
    g.fig.colorbar(
        g.ax_heatmap.collections[0], cax=cbar_ax,
        label="Mean F1",
    )
    cbar_ax.yaxis.label.set_size(11)
    cbar_ax.tick_params(labelsize=9)

    outpath = os.path.join(outdir, f"{key}_param_heatmap.png")
    g.savefig(outpath, bbox_inches="tight", dpi=200)
    plt.close(g.fig)
    print(f"  Saved {outpath} ({n_rows} combos × {n_cols} cell types)")


def save_legend(ref_short, ref_palette, sub_palette, outdir):
    """Save a standalone legend figure for the heatmap annotations."""
    fig, ax = plt.subplots(figsize=(6, 5))
    ax.axis("off")

    handles = []

    # Reference
    handles.append(mpatches.Patch(color="none", label="Reference"))
    for ref in sorted(ref_short.keys()):
        letter = ref_short[ref]
        handles.append(mpatches.Patch(color=ref_palette[ref],
                                      label=f"  {letter}: {ref}"))

    # Method
    handles.append(mpatches.Patch(color="none", label=""))
    handles.append(mpatches.Patch(color="none", label="Method"))
    for method in sorted(METHOD_COLORS.keys()):
        display = METHOD_NAMES.get(method, method)
        handles.append(mpatches.Patch(color=METHOD_COLORS[method],
                                      label=f"  {display}"))

    # Subsample
    handles.append(mpatches.Patch(color="none", label=""))
    handles.append(mpatches.Patch(color="none", label="Subsample Ref"))
    for sub in sorted(sub_palette.keys()):
        handles.append(mpatches.Patch(color=sub_palette[sub], label=f"  {sub}"))

    ax.legend(
        handles=handles, loc="center", fontsize=20,
        frameon=False, handlelength=1.5, handleheight=1.2,
    )

    fig.tight_layout()
    outpath = os.path.join(outdir, "param_heatmap_legend.png")
    fig.savefig(outpath, bbox_inches="tight", dpi=200)
    plt.close(fig)
    print(f"Saved legend: {outpath}")


def main():
    args = parse_args()
    os.makedirs(args.outdir, exist_ok=True)
    set_pub_style()

    df = pd.read_csv(args.input, sep="\t")
    print(f"Loaded {len(df)} rows from {args.input}")

    # --- Global color palettes ---
    references = sorted(df["reference"].unique())
    ref_short = build_ref_short_names(references)

    ref_palette = dict(zip(references, sns.color_palette("Set2", len(references))))

    subsample_refs = sorted(df["subsample_ref"].unique())
    sub_palette = dict(zip(subsample_refs, sns.color_palette("YlOrRd_d", len(subsample_refs))))

    # --- Standalone legend ---
    save_legend(ref_short, ref_palette, sub_palette, args.outdir)

    # --- One heatmap per key level ---
    keys = [k for k in KEY_ORDER if k in df["key"].unique()]
    for key in keys:
        print(f"\n{key}:")
        key_df = df[df["key"] == key]
        make_clustermap(key, key_df, ref_short, ref_palette, sub_palette, args.outdir)

    print(f"\nDone. Figures in {args.outdir}/")


if __name__ == "__main__":
    main()
